### DISCUSSION TITLE
Add Apple Music Support

### DISCUSSION THREAD

**Discussion by r3draid3r04** [music-assistant/support#389] [Feature requests and ideas]
# EDIT : Apple Music is now available https://music-assistant.io/music-providers/apple-music/

Watched the announcement and it was mentioned that apple music might not be possible due to api. I don't know if the apple music api can be integrated but I have seen some people use to api to make web apps the api documentation I have found is below.

https://developer.apple.com/documentation/musickitjs
https://developer.apple.com/documentation/applemusicapi/

Don't know if it is possible to do or not as I am not strong in this area but it would be cool. I also do not know if paying and becoming an apple dev is required either.

**Comment by jacobkaad01** [comment]
I was asked to repost this here.

"Does anyone know how Cider is streaming from Apple Music? Maybe that could be an avenue. https://github.com/ciderapp/Cider"

**Reply by timwnl** [reply]
In addition, it is a bit out dated but Musish created a whole Apple Music player. Maybe some code can be reused to give it a kick-start?

https://github.com/Musish/Musish

**Comment by hiddevanbrussel** [comment]
Would be nice if this is possible.

**Comment by AlexLamond** [comment]
Are there any plans to investigate how Cider/Musish added AM support and add it to Music Assistant?

**Comment by bgoncal** [comment]
So looking forward to it!

**Comment by JeffersonBledsoe** [comment]
Just a heads up that Apple MusicKit has a web beta with a fairly well designed (if constrained) API:
https://js-cdn.music.apple.com/musickit/v3/docs/index.html?path=/story/introduction--page

I've not been able to find a way to integrate music-assistant into my home yet so don't have interest in making a PR for this currently, but I'm sure somebody else could give it a go!

**Comment by KlasFNilsson** [comment]
Just added Music assistant and it’s great!! Kinda feels like I should leave Apple Music for Spotify, just for this. Or could we hope for a solution in near time?.. 😇

**Reply by OzGav** [reply]
Still waiting on a dev to pick this up I’m afraid. All existing devs are fully occupied with existing functionality

**Reply by KlasFNilsson** [reply]
Ah. Thanks for the quick reply, though!

**Reply by HA-TB303** [reply]
Hope there will be more priority on this. For apple music there is literally NOTHING to get connected to HA, so Music Assistant is my only hope. ;)

**Reply by domemshur** [reply]
Any word on this?

**Reply by OzGav** [reply]
Still waiting for a volunteer

**Reply by domemshur** [reply]
Are there any devs that can look into this?

**Reply by OzGav** [reply]
Who are you asking with that question?

**Comment by OptimusGREEN** [comment]
Yeah I'd really like apple music support too. I hope this comes to fruition.

**Comment by domemshur** [comment]
Is anyone working on an Apple Music integration?

**Reply by OzGav** [reply]
No. Are you volunteering?

**Reply by HA-TB303** [reply]
Doesn’t look this is getting anywhere soon as the music assistant devs don’t seem to be interested much in apple music. For a music assistant to assist playing music, it first needs to be able to play the music ;)

**Reply by marcelveldt** [reply]
We are busy enough with other stuff atm, writing a new music provider does not have much priority. Somebody could pick up the glove and write one, many devs have followed that example. There does not seem to be enough interest in Apple Music at this time. I'm pretty sure one of ill probably take a look after the 2.0 release if nobody volunteers.

**Reply by HA-TB303** [reply]
I beg to differ: https://github.com/orgs/music-assistant/discussions/1169
![CleanShot 2024-02-29 at 09 22 47](https://github.com/music-assistant/hass-music-assistant/assets/9609805/66c4fbda-ab13-4023-82e6-8e899cc74f1c)


Anyway, we'll have to wait, which is a shame since Music Assistant is another great product I can not use.

**Reply by OzGav** [reply]
Let me add something to Marcel’s sentence to make it clearer. There does not seem to be enough interest in Apple Music _**for someone to volunteer their time to write the provider**_. We don’t need people being argumentative and ungrateful.

**Reply by HA-TB303** [reply]
I am in no way ungrateful, I am just expressing my disappointment for not being able to use Music Assistant.

**Reply by valerix85** [reply]
> We are busy enough with other stuff atm, writing a new music provider does not have much priority. Somebody could pick up the glove and write one, many devs have followed that example. There does not seem to be enough interest in Apple Music at this time. I'm pretty sure one of ill probably take a look after the 2.0 release if nobody volunteers.

If it's okay with you I can try I would need to know what guidelines to follow though so that I am in line with the project. Thank you

**Reply by OzGav** [reply]
Please join discord and ask questions in the devs-music providers channel. You can just review the existing provider code and mimic that.

**Reply by valerix85** [reply]
Thanks @OzGav I will join today after work

**Reply by OzGav** [reply]
This is the dev docs at the moment https://github.com/music-assistant/server/blob/main/DEVELOPMENT.md#-building-your-own-music-provider

**Reply by mjkonko** [reply]
Someone correct me please if I'm wrong but isn't an active Apple Developer Program license required to get the tokens to even start working with the Music API? I'm kinda into giving a shot and implementing the new provider that but neither do I have an active license anymore nor a spare 100 bucks laying around.

_Edit: confirmed, got a nice warning when trying to get into the Certificates of developer.apple.com_

**Reply by OzGav** [reply]
Anyone wishing to contribute to MA is strongly encouraged to seek assistance on Discord. Developer Program assistance may be available.

**Reply by Hackashaq666** [reply]
I don't know if this is proof of concept or not, but there is a project that recently became open source that enables access to  Apple music and a few other services.  I have no connection to the project.

https://github.com/rekkyrosso/ampcast

**Reply by OzGav** [reply]
Apple Music was added to the latest beta yesterday

**Reply by domemshur** [reply]
I saw that. It looks great. I am have no issues getting my playlists and songs though. Only Artist and albums show up.

**Reply by OzGav** [reply]
Can you reword that as it doesn’t make sense to me currently.

**Comment by domemshur** [comment]
Only my artists and albums are showing up. Not my songs or playlists.

**Reply by OzGav** [reply]
@MarvinSchenkel

**Reply by MarvinSchenkel** [reply]
Anything interesting in the logs?

**Reply by domemshur** [reply]
<img width="1159" alt="1" src="https://github.com/music-assistant/hass-music-assistant/assets/81906007/562363c5-c2bf-4470-a672-eb4d5dd6b8d6">

**Reply by domemshur** [reply]
I found this in the settings and when I clicked it I got this error

<img width="1159" alt="1" src="https://github.com/music-assistant/hass-music-assistant/assets/81906007/2eee0dd4-d407-4460-9da7-56ffeed28bee">
<img width="1159" alt="2" src="https://github.com/music-assistant/hass-music-assistant/assets/81906007/4e85623f-24d4-4bca-b903-14202e7fe362">

**Reply by OzGav** [reply]
What settings are you referring to?

**Reply by MarvinSchenkel** [reply]
> <img alt="1" width="1159" src="https://private-user-images.githubusercontent.com/81906007/333296597-562363c5-c2bf-4470-a672-eb4d5dd6b8d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY1MzYyNTMsIm5iZiI6MTcxNjUzNTk1MywicGF0aCI6Ii84MTkwNjAwNy8zMzMyOTY1OTctNTYyMzYzYzUtYzJiZi00NDcwLWE2NzItZWI0ZDVkZDZiOGQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI0VDA3MzIzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ4MWUyYWRlYzE4NjQzZjI5ZWM2ZDZiYWU1Y2NlYjQ0NDhhNzVhNmNlZGM4YzJmMzMxYjNkYzRmOWNlNWM5ZDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.BDVPE-e3hM1QBqh0dg7eeImPcAXFPSHYlhmK0er3vOI">

Added a fix for this error, so keep an eye out for the next beta.

**Comment by domemshur** [comment]
[
![IMG_5732](https://github.com/music-assistant/hass-music-assistant/assets/81906007/0d54f920-03d6-4731-87c6-61b3a1507e5b)
![IMG_5733](https://github.com/music-assistant/hass-music-assistant/assets/81906007/342e06d1-6111-436d-b6a4-e37c7b22033b)
](url)

**Comment by nicopret1** [comment]
Not sure what I'm missing, but I don't see the option to add Apple Music if I click "Add Music Provider?
![Screenshot 2024-06-09 at 13 47 01](https://github.com/music-assistant/hass-music-assistant/assets/130547447/2de164b0-0c42-434e-89a9-bc3ba3469102)

**Reply by nicopret1** [reply]
After looking at the log I realised that although I was running the Beta Addon and it seemed to be working it was not the case. Both the Beta and "normal" (release) versions were running and the Beta interface version simply returned the same data as the "normal" (release) version. You need to stop the "normal" version of the Addon since it seems only one of the Addon's can connect to the streaming server at a time. Check the log to confirm all is working. If you then go to "Add Music Provider" the Apple Music option is available and you can add it as per the documentation. I also used Safari to retrieve the token rather than Chrome (which for some reason was terribly slow when trying to login to Apple Music on my Mac).

**Comment by xmirakulix** [comment]
@MarvinSchenkel First off, thank you for your work, I highly appreciate your effort!
Works nicely, but I ran into an error and opened an issue for it: https://github.com/music-assistant/hass-music-assistant/issues/2431#issue-2342424947

**Reply by MarvinSchenkel** [reply]
Thanks for reporting!

**Comment by Kebo2r** [comment]
I'm running beta 8 and having the issue with only artists and albums showing.  No tracks or playlists.  Should this be the case in B8?  Thank you!

**Reply by Kebo2r** [reply]
> I'm running beta 8 and having the issue with only artists and albums showing. No tracks or playlists. Should this be the case in B8? Thank you!

Here's the debug log:

2024-06-28 16:25:52.057 WARNING (MainThread) [music_assistant] Exception in task Task-66 - target: <coroutine object MusicController._start_provider_sync.<locals>.run_sync at 0x7f7c969f30>: 500, message='Internal Server Error', url=URL('https://api.music.apple.com/v1/me/library/albums?include=catalog,artists&extend=editorialNotes&limit=50&offset=50')
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/music_assistant/server/controllers/music.py", line 700, in run_sync
    await provider.sync_library(media_types)
  File "/usr/local/lib/python3.12/site-packages/music_assistant/server/models/music_provider.py", line 396, in sync_library
    async for prov_item in self._get_library_gen(media_type):
  File "/usr/local/lib/python3.12/site-packages/music_assistant/server/providers/apple_music/__init__.py", line 192, in get_library_albums
    for item in await self._get_all_items(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/music_assistant/server/providers/apple_music/__init__.py", line 583, in _get_all_items
    result = await self._get_data(endpoint, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/music_assistant/server/helpers/throttle_retry.py", line 72, in wrapper
    return await func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/music_assistant/server/providers/apple_music/__init__.py", line 615, in _get_data
    response.raise_for_status()
  File "/usr/local/lib/python3.12/site-packages/aiohttp/client_reqrep.py", line 1070, in raise_for_status
    raise ClientResponseError(
aiohttp.client_exceptions.ClientResponseError: 500, message='Internal Server Error', url=URL('https://api.music.apple.com/v1/me/library/albums?include=catalog,artists&extend=editorialNotes&limit=50&offset=50')
2024-06-28 16:25:52.060 WARNING (MainThread) [music_assistant.music] Sync task for Apple Music completed with errors
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/music_assistant/server/controllers/music.py", line 700, in run_sync
    await provider.sync_library(media_types)
  File "/usr/local/lib/python3.12/site-packages/music_assistant/server/models/music_provider.py", line 396, in sync_library
    async for prov_item in self._get_library_gen(media_type):
  File "/usr/local/lib/python3.12/site-packages/music_assistant/server/providers/apple_music/__init__.py", line 192, in get_library_albums
    for item in await self._get_all_items(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/music_assistant/server/providers/apple_music/__init__.py", line 583, in _get_all_items
    result = await self._get_data(endpoint, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/music_assistant/server/helpers/throttle_retry.py", line 72, in wrapper
    return await func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/music_assistant/server/providers/apple_music/__init__.py", line 615, in _get_data
    response.raise_for_status()
  File "/usr/local/lib/python3.12/site-packages/aiohttp/client_reqrep.py", line 1070, in raise_for_status
    raise ClientResponseError(
aiohttp.client_exceptions.ClientResponseError: 500, message='Internal Server Error', url=URL('https://api.music.apple.com/v1/me/library/albums?include=catalog,artists&extend=editorialNotes&limit=50&offset=50')

**Reply by MarvinSchenkel** [reply]
Hm this is strange.. This might be due to an expired token.. Can you remove the provider and try the login process again? I do have some other fixes coming up in the next beta as well

**Comment by justinwalsh** [comment]
Apple has released an official authorization flow for Apple Music integrations. I looked around the issues and ideas boards and didn't see it mentioned yet, so I thought I would share the news.

https://js-cdn.music.apple.com/musickit/v3/docs/index.html?path=/story/user-authorization--page

**Reply by MarvinSchenkel** [reply]
I played around with this and unfortunately this only supports Javascript. Music assistant backend is written in python, so this won't work I'm afraid

**Reply by jan-auer** [reply]
It looks like there's an undocumented way to extract a Music User Token from the Web API. Has someone looked into bringing up a screen in the frontend loaded with MusicKit to perform authentication and then pass the token on to the backend?

Not sure that would be as straightforward within HASS, but I could spend a little time on this over Christmas if this sounds promising.

**Reply by marcelveldt** [reply]
All help is welcome! I'm sure many people will be very thankful if you manage to sort it out. Maybe join our discord server to discuss progress with @MarvinSchenkel ?

**Reply by MarvinSchenkel** [reply]
Worth having a shot. I did have a look at it, but  because it's all client side, we wouldn't be able to automatically refresh the token. Then again, it's a much more user friendly approach if you get this to work even without an auto refresh.. Feel free to ask me for any help on discord if you're stuck with something 💪

**Reply by finnlehmkuhl** [reply]
maybe take a look at
https://developer.apple.com/documentation/storekit/offering-apple-music-subscription-in-your-app

> ... unfortunately this only supports Javascript

Look to me like it is using swift or Objective-C

Maybe combined with https://pyobjc.readthedocs.io/en/latest/

PyObjC provides bindings to most Objective-C frameworks ... build upon a generic bidirectional bridge between Python and Objective-C

I am not a developer and I am sorry if I am totally of track here - just wants my Apple Music available in HA.
I have only just begun this journey, but I am looking forward to enjoy  Music Assistant in HA.

**Comment by NRutt2005** [comment]
I am trying to add Apple Music as a music provider within my Music Assistant. I followed the instructions on the https://www.music-assistant.io/music-providers/apple-music/#features documentation. When I follow these steps, I get a 401, message='Unauthorized' error. (See screenshot).

Has anyone ran into this issue and were you able to solve the issue? Thank you

![Screenshot 2025-04-25 110151](https://github.com/user-attachments/assets/14b1a456-0446-4e47-b47c-bc1786fd77d2)

**Reply by spencer-scott** [reply]
It's something to do with their Apple Developer account. See the [Discord discussion](https://discord.com/channels/753947050995089438/1362539256894984366 )

**Reply by SSch1987** [reply]
Hey,
I just had the same issue. Tried to re-add the music provider. However, I have no clue about how to get a "MusicKit App Token". This is not the User Token as described in the documentation. Can anyone help?

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
